### PR TITLE
Use MaIS APIs to get course term and individual course information

### PR DESCRIPTION
**Pull request by hudajkhan**
Description of what pull request does:

The original code relied on the fetch_xml rake task to retrieve the XML for course terms (e.g. Spring 2023, Summer 2023, etc.) generated by the registry harvester. The rake task copied the XML to a local directory (course_work_xml). When the user clicked on the reserves list button, an AJAX request used the "all_courses" method in "course_work_courses" which loaded the XML files, parsed them, retrieved course, section, and instructor information, saved that info in Course (a Ruby class) objects, and returned those objects to the requesting call.

The updates in this pull request are no longer relying on the XML files that are copied over from registry harvester's output, but are using the MaIS course term API request to get the list of courses for a given course term (e.g. spring 2023, summer 2023), and then using the course API to get individual course information for sections and their instructors. In addition, we're parsing the XML from these API requests, but saving the information as JSON files to be read in by the app when the user clicks the button to retrieve the list of courses. These JSON files save only the information required by the Course objects. Each JSON file representing a course term has a list of Course JSON objects which mirror the info required by the Ruby Course objects used in the app.

The steps that occur now are:

- Fetch_courses: Fetch_xml has been renamed to fetch_courses, and instead of copying the XML file from registry harvester, calls the Course Term MaIS API to get course term information, and then uses a new class called CourseAPI class to parse the course term info returned, retrieve section and instructor information, and generate the JSON files that will sit in course_work_xml to be read in when the user clicks the button to get all the reserves info. The process of generating these JSON files takes a few hours, so this step is better left as a rake task that can run independently and then make the info available to the app. The result is a JSON file per term saved in the "course_work_xml" directory. The original code requested the current and next immediate term, so that pattern remains the same.
- When the user clicks the button to look at the list of courses, they trigger the "all_courses" method in Course_work_courses. This class has replaced the initialization step of loading XML to loading the JSON files from course_work_xml instead, and replaced the processing step that looped through the registry harvester course term XML to pick out only those sections with instructors, with a simpler step of reading in the JSON to generate the Course objects.

In addition, we've added some test files and updated the existing course_work_courses spec. The API cert and key files are saved in the shared configs directory (and would be saved in the config directory when running the app locally), so *.cert and *.key files under config have been added to .gitignore.

### CODE REVIEW

File: config/settings.yml [hunk 1]
```diff
@@ -248,3 +248,5 @@ terms:
 - term: Summer 2027
   quarter: Summer
   end_date: 2027-08-15
+
+certs_path: "../../shared/config/"
```
Comment (jcoyne) [inline config/settings.yml:252]: In the app, this gets linked to `Rails.root.join('config')`. There shouldn't be a need to make this a special setting.

File: lib/course_api.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        client_cert = OpenSSL::X509::Certificate.new File.read(cert_file)
+        client_key = OpenSSL::PKey.read File.read(key_file)
+        connection = Faraday::Connection.new 'https://registry.stanford.edu', :ssl => { :client_cert => client_cert, :client_key => client_key }
+        spec_course = '/doc/courseclass/' + request_class_id
+        begin
+            response = connection.get spec_course
+        # Timeout error
+        rescue Faraday::TimeoutError
+            errors << "timeout error occurred for first attempt for " + request_class_id.to_s
+            # Retry
```
Comment (jcoyne) [inline lib/course_api.rb:72]: There is a faraday-retry gem that can handle this for us: https://github.com/lostisland/faraday-retry

File: lib/course_api.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                errors << "retry failed for " + request_class_id.to_s
+            end
+        # Connection failed error
+        rescue Faraday::ConnectionFailed
+            errors << "Connection failed occurred for first attempt for " + request_class_id.to_s
+        rescue Faraday::ServerError
+            errors << "Server Error for " + request_class_id.to_s
+        end
+        if response.status == 200
+            course_response = Nokogiri::XML(response.body)
```
Comment (jcoyne) [inline lib/course_api.rb:85]: Can this XML parsing be extracted to a separate method or class? That would make the code easier to reason about.

File: lib/course_api.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        if(term_quarter == "Fall") then academic_year += 1 end
+        if(term_id[0,1] == "1") then academic_year = "20" + academic_year.to_s end
+        term_display = term_quarter + " " + academic_year.to_s 
+    end
+
+    def get_term_quarter(term_id)
+        term_suffix = term_id[3,3]
+
+        #Get last digit
+        case term_suffix
```
Comment (jcoyne) [inline lib/course_api.rb:124]: Can this be a constant hash? 
```
TERM_QUARTER = { '2' => 'Fall', '4' => 'Winter', '6' => 'Spring', '8'=>'Summer' }
```

Comment (hudajkhan) [reply lib/course_api.rb:124]: Sounds good. Thanks

**Comment by hudajkhan** [comment]
Description of what pull request does:

The original code relied on the fetch_xml rake task to retrieve the XML for course terms (e.g. Spring 2023, Summer 2023, etc.) generated by the registry harvester.  The rake task copied the XML to a local directory (course_work_xml).  When the user clicked on the reserves list button, an AJAX request used the "all_courses" method in "course_work_courses" which loaded the XML files, parsed then, retrieved course, section, and instructor information,  saved that info in Course objects, and returned those objects to the requesting call.  

The updates in this pull request are no longer relying on the XML files that are copied over from registry harvester's output, but are using the MaIS course term API request to get the list of courses for a given course term (e.g. spring 2023, summer 2023), and then using the course API to get individual course information for sections and their instructors. In addition, we're parsing the XML from these API requests, but saving the information as JSON files to be read in by the app when the user clicks the button to retrieve the list of courses.  These JSON files save only the information required by the Course objects.  Each JSON file representing a course term has a list of Course JSON objects which mirror the info required by the Ruby Course objects used in the app. 

The steps that occur now are:

- Fetch_courses: Fetch_xml has been renamed to fetch_courses, and instead of copying the XML file from registry harvester, calls the Course Term MaIS API to get course term information, and then uses a new class called CourseAPI class to parse the course term info returned, retrieve section and instructor information, and generate the JSON files that will sit in course_work_xml to be read in when the user clicks the button to get all the reserves info. The process of generating these JSON files takes a few hours, so this step is better left as a rake task that can run independently and then make the info available to the app. The result is a JSON file per term saved in the "course_work_xml" directory.  The original code requested the current and next immediate term, so that pattern remains the same.  
- When the user clicks the button to look at the list of courses, they trigger the "all_courses" method in Course_work_courses. This class has replaced the initialization step of loading XML to loading the JSON files from course_work_xml instead, and replaced the processing step that looped through the  registry harvester course term XML to pick out only those sections with instructors, with a simpler step of reading in the JSON to generate the Course objects. 

In addition, we've added some test files and updated the existing course_work_courses spec.  The API cert and key files are saved in the shared configs directory (and would be saved in the config directory when running the app locally), so *.cert and *.key files under config have been added to .gitignore.

File: lib/course_api.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        @course_term = Nokogiri::XML(xml)
+    end
+    # Parse the XML returned by the CourseTerm MaIS API
+    def parse
+        courses = []
+        all_errors = []
+        # Return array of course hash objects
+        courses_list = parse_courses()
+        counter = 0
+        courses_list.each do |course|
```
Comment (cbeer) [inline lib/course_api.rb:18]: Could we do away with the internal counter if we use `each_slice` instead?

```
courses_list.each_slice(10000) do |courses|
  courses.each do |course|
     ...
  end

  sleep(1)
end
```

or, I suppose, use each_with_index and `index % 10000`?

File: lib/course_api.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                counter = 0
+            end 
+            request_class_id = course[:request_class_id]
+            course_info = get_course_info(request_class_id)
+            sections = course_info[:sections]
+            all_errors.concat(course_info[:errors]) 
+            sections.each do |section|
+                section_id = section[:sid]
+                instructors = section[:instructors]
+                course_hash = {"title": course[:title], "term": course[:term], "cid": course[:cid], "cids": course[:cids],
```
Comment (cbeer) [inline lib/course_api.rb:31]: I'm assuming string-type keys are necessary for compatibility with the old behavior or database serialization, or could we use `with_indifferent_access` and 🙈 ? The mix of string and symbol keys makes it a little hard to follow.

File: lib/course_api.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+                name = person.text
+                instructors << {"sunet": sunetid, "name": name}
+            end
+            sections << {"sid": section_id, "instructors": instructors}
+        end
+        sections
+    end
+
+    # Based on logic https://github.com/sul-dlss/registry-harvester/blob/main/Course/src/main/java/edu/stanford/BuildTermString.java
+    def generate_term(term_id)
```
Comment (cbeer) [inline lib/course_api.rb:132]: Could we document the term_id format in the method comment? I think I get it, but...

File: lib/tasks/fetch_courses.rake [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  errors = []
+  updated = false
+
+  # Use environment variable with a default value set to a relative path
+  # to identify where this file should be
+  cert_file = Rails.root.join("config", "sul-harvester.cert")
+  key_file = Rails.root.join("config","sul-harvester.key")
+  client_cert = OpenSSL::X509::Certificate.new File.read(cert_file)
+  client_key = OpenSSL::PKey.read File.read(key_file)
+  connection = Faraday::Connection.new 'https://registry.stanford.edu', :ssl => { :client_cert => client_cert, :client_key => client_key }
```
Comment (cbeer) [inline lib/tasks/fetch_courses.rake:19]: Maybe we could pull connection up to e.g. `CourseAPI.connection` (and share it with the `CourseAPI#request_course_api` connection stuff too)?

**Comment by hudajkhan** [comment]
I'm closing for now since I'm be setting up new pull requests. Thanks for your suggestions! I've tried to work them in!
